### ISSUE TITLE
Upgrade to Unity 2019.4.40 from 2019.4.19

### ISSUE THREAD

**Issue by Frooxius** [Yellow-Dog-Man/Resonite-Issues#585]
### Is your feature request related to a problem? Please describe.

Currently we're using older version of Unity 2019.4.19, which has some issues. We'd want to upgrade at least to the latest minor version, which is 2019.4.40, which should have additional bugfixes.

### Describe the solution you'd like

Upgrade to 2019.4.40 for the builds and resolve issues. Last time upgrade was attempted, the bloom filter freaks out in VR, which is something that needs to be diagnosed and addressed. 

We likely do not want to spend the engineering effort to upgrade to a new major version of Unity, since we're overall moving to a custom rendering engine instead, but upgrading the minor version in the interim can be beneficial to potentially address some issues for now.

### Describe alternatives you've considered

We can stay on 2019.4.19 until we switch to a custom rendering engine. Alternatively upgrade to a new major version, but that's significantly bigger undertaking, with more engineering challenges to resolve.

### Additional Context

This might help resolve some crashes and issues, such as: https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/269

**Comment by Geenz** [comment]
Biggest problem on this one is we'd need to use multi-pass VR rendering, which does incur overhead.  Reason for this is Unity's post processing stack doesn't support single pass rendering in the deferred pipeline in any shape or form.  This would incur a performance penalty - but I'm not sure how acceptable this would be without circulating some builds to test.

**Comment by Frooxius** [comment]
@Geenz I do not understand your comment.

- In Unity 2019.4.19 which we are using right now, we are using deferred pipeline with single pass rendering, which works with the post processing stack
- In the Unity 2019.4.40 build, the post processing stack still worked with deferred + single pass rendering, but the bloom filter has a bug of not clearing up properly and blooming from the edges, which I asked to be investigated and fixed

Are you suggesting that between the minor version 2019.4.19 and 2019.4.40, Unity has dropped support for this?

You suggest that there's no support in any shape or form, but except for the specific bug with bloom, the post processing still seems to work as it does in 2019.4.19. Is this blooming bug fundamentally not fixable at all? If so, can you point me to specific feature/change that Unity did in those minor versions that made this not fixable?

**Comment by Geenz** [comment]
@Frooxius I noticed it wasn't just the bloom filter causing this.

More generally, *something* in the post process stack will cause progressive over exposure with single pass rendering in the deferred renderer.  But only single pass - multi pass works fine.  This is after doing a few builds to A/B test this.  It seems like it's something deeper in the post process stack that's creating this, and this is both with an updated version of the post process stack and the one we're presently using in 2019.4.19.

I'm not sure what specifically changed between 19 and onwards, but it reproduces from 2019.4.20 onwards from my testing going through multiple versions inbetween to see if it was just a specific quirk that I was missing inbetween.

**Comment by Frooxius** [comment]
@Geenz That sounds like it is single-pass specific bug, but nothing about that seems to indicate to me that the post processing stack + single pass deferred would be fundamentally incompatible in 2019.4.40 and that whatever is the root cause in this would not be fixable.

It seems odd to me that something between 2019.4.19 and 2019.4.40 would change to fundamentally that this would not be a fixable issue. Are you able to investigate this deeper to find the root cause?

**Comment by Geenz** [comment]
@Frooxius It starts with 2019.4.20 - and I haven't been able to more deeply investigate since I last touched this.  I will note what I've done thus far though:

Single-pass:
- Disable bloom on 4.20, 4.40, and a few Unity versions in between
- Disable color grading on 4.20, 4.40

It had the same bug where the screen was not seemingly clearing correctly in both cases.

Multi-pass:
- Nothing needed to be disabled for this to "just work"
  - I didn't notice any performance impact on my machine, however my machines are far from indicative of what our average user will be using (a threadripper and a CPU with V-Cache are not exactly the best judges here)

To me this indicates something deeper in the post processing stack for single pass.  I did look a little bit, and the code _generally_ checked out.  Newer versions of the post processing stack also exhibits this - so either:
A. It's a bug that got introduced after 4.19 with the single pass on deferred pipeline specifically
B. We're doing something that's making the post processing stack upset (which is weird since we don't really do anything particularly special here).

**Comment by shiftyscales** [comment]
> I haven't been able to more deeply investigate since I last touched this.

Have you looked into this issue at all since your last update, @Geenz?

Moreover- do we still want to prioritize this minor version bump, or would it be better if we ride things out until we can get off of Unity entirely, @Frooxius?

For context- I'm going through our list of "blocked" issues- but don't explicitly see anything that is blocking this issue aside from investing time into looking into the post-processing issues it reportedly causes- so I've adjusted its status accordingly.

**Comment by Frooxius** [comment]
@shiftyscales Given that there are now even more issues popping up with Unity crashes (e.g. with SteamLink), I think it makes this more important to do.

It will still take some time to switch another rendering engine, so we'd have to deal with those issues until then.

And yes, I don't see anything blocking this issue - other than just working on it to resolve the problem with the rendering.

**Comment by Nytra** [comment]
Bloom bug / frame not clearing bug sounds similar to this issue https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/2847

**Comment by Frooxius** [comment]
@Nytra It's a different issue.

- This happens *only* in VR, while #2847 happens in desktop.
- This happens specifically after upgrade to the Unity version, not with current build
- This also doesn't turn everything to white, just around objects

**Comment by Frooxius** [comment]
I've made a build with 2019.4.40f internally to test if it helps with some of these issues - while the build has rendering issue while in VR, most of the stuff works enough to be able to test the issue.

I've got unfortunate news though. I've tested both the forceMatrixRecalculationPerRender and the controller swapping issues with SteamLink and in both cases I've reproduced the crash even with 2019.4.40f

This is log for controller swapping: 
[Player.log](https://github.com/user-attachments/files/17563759/Player.log)
[FROOXIAN - 2024.10.29.1229 - 2024-10-29 16_14_50.log](https://github.com/user-attachments/files/17563773/FROOXIAN.-.2024.10.29.1229.-.2024-10-29.16_14_50.log)

And this is for forceMatrixRecalculationPerRender:
[Player.log](https://github.com/user-attachments/files/17563764/Player.log)
[FROOXIAN - 2024.10.29.1229 - 2024-10-29 16_22_06.log](https://github.com/user-attachments/files/17563774/FROOXIAN.-.2024.10.29.1229.-.2024-10-29.16_22_06.log)

This means that this upgrade is likely not worth doing at this point, since it can't resolve either of those issues and we'd need to fix the rendering issue in order to perform the upgrade anyway.

The other unfortunate news is that we also don't have a clear easy path to fixing either of those problems in the interim now (actually... there's one way to fix issue with RTO - we actually duplicate the mesh renderers and use masks to select the correct ones to render, but it'll require some rework of RTO).

But for the upgrade itself, I'd say we nix this, since it won't help with those and we focus on getting away from Unity as fast as we can.
